### DISCUSSION TITLE
Using the 12864-06D display, and how does flipmode work?

### DISCUSSION THREAD

**Discussion by RvanDeelen** [olikraus/u8g2#1791] [General]
So a while ago i bought some displays of aliexpress. As one does. They were explicitly specified as being "right to left written". 
It took me ages to get t the right constructor but in the end i found out that  "U8G2_ST7565_NHD_C12864_1_4W_SW_SPI" will work. Because before this i used U8G2_ST7565_KS0713_1_4W_SW_SPI constructor, but altered the ADC reversal line manually in the library.
So now everything is oriented correctly (writing left to right and no mirrored letters), For my use case i did have to use rotation R2 but that's besides the point.

Something i don't understand tho is how the flipmode works in the constructor code. how can i know if the flipmode instruction actually works looking at the library?

**Comment by olikraus** [comment]
The code for flipmode is only sent to the display if the flipmode command (https://github.com/olikraus/u8g2/wiki/u8g2reference#setflipmode) is used. Ideally setFlipMode(0) should be executed automatically by the constructor. Unfortunately this isn't done (I missed this very early in the development now I am too lazy to fix this).
As a result, the commands for flip mode 0 must be implemented in the constructor startup code and in the flipmode 0 sequence itself.

In your case:
https://github.com/olikraus/u8g2/blob/d9402d5e40ca1224a300c39d9baef4051eefb7b5/csrc/u8x8_d_st7565.c#L804

The flip mode sequence is sent here:
https://github.com/olikraus/u8g2/blob/d9402d5e40ca1224a300c39d9baef4051eefb7b5/csrc/u8x8_d_st7565.c#L819-L830

And the commands are repeated in the normal startup sequence of that display:
https://github.com/olikraus/u8g2/blob/d9402d5e40ca1224a300c39d9baef4051eefb7b5/csrc/u8x8_d_st7565.c#L787-L802

Which is actually defined here:
https://github.com/olikraus/u8g2/blob/d9402d5e40ca1224a300c39d9baef4051eefb7b5/csrc/u8x8_d_st7565.c#L59-L73

But it seems, that the flip mode sequences are wrong. and the code for flip mode 0 and startup do not match.
In startup we have the commands 0xa0 and 0xc0, however flipmode 0 will send the wrong code 0xa1 0xc0.
This means, the display will be ok as long as you do not use the flipmode command, but will destroy the content after first use of flipmode.

Is this the problem you observed?

To fix this, I need to change
https://github.com/olikraus/u8g2/blob/d9402d5e40ca1224a300c39d9baef4051eefb7b5/csrc/u8x8_d_st7565.c#L822
to zflip1 and
https://github.com/olikraus/u8g2/blob/d9402d5e40ca1224a300c39d9baef4051eefb7b5/csrc/u8x8_d_st7565.c#L827
to zflip0

hmm looks like a bug in u8g2...

**Comment by olikraus** [comment]
Created issue https://github.com/olikraus/u8g2/issues/1792 for this.
Lets continue the discussion there.
